### PR TITLE
feat: added backer as a new lender adapter

### PR DESCRIPTION
**Pull request by manuelpolzhofer**

### CODE REVIEW

File: src/backer.sol [hunk 1]
```diff
@@ -0,0 +1,56 @@
+pragma solidity >=0.4.24;
+
+contract LenderTokenLike {
```
Comment (lucasvo) [inline src/backer.sol]: Isn't this enough:

```suggestion
contract TokenLike {
```

File: .gitmodules [hunk 2]
```diff
@@ -13,3 +13,6 @@
 [submodule "lib/dss-add-ilk-spell"]
 	path = lib/dss-add-ilk-spell
 	url = https://github.com/makerdao/dss-add-ilk-spell
+[submodule "lib/ds-token"]
```
Comment (lucasvo) [inline .gitmodules]: Seems to be unused?

File: src/backer.sol [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    function provide(address usrC, address usrT, uint wadC, uint wadT) auth public {
+        require(tkn.balanceOf(backer)>= wadT);
+        collateral.transferFrom(usrC, backer, wadC);
+        tkn.transferFrom(backer,usrT, wadT);
+
+    }
+
+    function release(address usrC, address usrT, uint wadC, uint wadT) auth  public {
+        require(collateral.balanceOf(backer)>= wadC);
+        tkn.transferFrom(usrT,backer, wadT);
```
Comment (lucasvo) [inline src/backer.sol]:
```suggestion
        tkn.transferFrom(usrT, backer, wadT);
```

File: src/backer.sol [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    mapping (address => uint) public wards;
+    function rely(address usr) public auth { wards[usr] = 1; }
+    function deny(address usr) public auth { wards[usr] = 0; }
+    modifier auth { require(wards[msg.sender] == 1); _; }
+
+    // --- Data ---
+    LenderTokenLike public tkn;
+    LenderTokenLike public collateral;
+
+    address private backer;
```
Comment (lucasvo) [inline src/backer.sol]: why `private`?

File: src/backer.sol [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+    }
+
+    function release(address usrC, address usrT, uint wadC, uint wadT) auth  public {
+        require(collateral.balanceOf(backer)>= wadC);
+        tkn.transferFrom(usrT,backer, wadT);
+        collateral.transferFrom(backer, usrC, wadC);
+    }
+
+    function setBacker(address usr) auth public {
```
Comment (lucasvo) [inline src/backer.sol]: I'd like to use: 
```suggestion
    function file(address usr) auth public {
```

File: src/backer.sol [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    function deny(address usr) public auth { wards[usr] = 0; }
+    modifier auth { require(wards[msg.sender] == 1); _; }
+
+    // --- Data ---
+    LenderTokenLike public tkn;
+    LenderTokenLike public collateral;
+
+    address private backer;
+
+    constructor (address tkn_, address collateral_, address backer_) public {
```
Comment (lucasvo) [inline src/backer.sol]: I think it's better not to require backer_ at instantiation. It makes the deploy more complicated. 
```suggestion
    constructor (address tkn_, address collateral_) public {
```

File: src/backer.sol [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    function transferFrom(address, address, uint) public;
+    function mint(address, uint) public;
+    function approve(address usr, uint wad) public returns (bool);
+    function balanceOf(address usr) public returns (uint);
+}
+
+contract BackerFab {
+    function deploy(address tkn_, address collateral_,address backer, address lightswitch_) public returns (address) {
+        Backer lender = new Backer(tkn_, collateral_,backer );
+        lender.rely(msg.sender);
```
Comment (lucasvo) [inline src/backer.sol]:
```suggestion
        lender.rely(msg.sender);
        lender.deny(this);
```

File: src/backer.sol [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        tkn = LenderTokenLike(tkn_);
+        collateral = LenderTokenLike(collateral_);
+        backer = backer_;
+    }
+
+    // --- Backer Methods ---
+    function provide(address usrC, address usrT, uint wadC, uint wadT) auth public {
+        require(tkn.balanceOf(backer)>= wadT);
+        collateral.transferFrom(usrC, backer, wadC);
+        tkn.transferFrom(backer,usrT, wadT);
```
Comment (lucasvo) [inline src/backer.sol]:
```suggestion
        tkn.transferFrom(backer, usrT, wadT);
```

File: src/backer.sol (hunk 4 above)
Comment (manuelpolzhofer) [reply src/backer.sol]: I know ever private variable is publicly visible but I thought it could be a bit more secure if it is not on chain accessible.

Comment (lucasvo) [reply src/backer.sol]: What would be more secure? It's less secure because you can't see who the current backer is.

File: src/backer.sol (hunk 5 above)
Comment (lucasvo) [reply src/backer.sol]: Also, let's use ds-note here and log the event.

**Pull request merged**